src/util/sighash.rs

```diff
@@ -0,0 +1,836 @@
+        {
+            let mut enc = TapLeafHash::engine();
+            leaf_version.consensus_encode(&mut enc)?;
+            script.consensus_encode(&mut enc)?;
+            let hash = TapLeafHash::from_engine(enc);
+
+            hash.into_inner().consensus_encode(&mut writer)?;
+            0u8.consensus_encode(&mut writer)?;
+            code_separator_pos.consensus_encode(&mut writer)?;
+        }
+
+        Ok(())
+    }
+
+    /// Compute the BIP341 sighash for any flag type.
+    pub fn taproot_signature_hash(
+        &mut self,
+        input_index: usize,
+        prevouts: &Prevouts,
+        annex: Option<Annex>,
+        script_path: Option<ScriptPath>,
+        sighash_type: SigHashType,
+    ) -> Result<TapSighashHash, Error> {
+        let mut enc = TapSighashHash::engine();
+        self.taproot_encode_signing_data_to(
+            &mut enc,
+            input_index,
+            prevouts,
+            annex,
+            script_path,
+            sighash_type,
+        )?;
+        Ok(TapSighashHash::from_engine(enc))
+    }
+
+    /// Encode the BIP143 signing data for any flag type into a given object implementing a
+    /// [std::io::Write] trait.
+    pub fn segwit_encode_signing_data_to<Write: io::Write>(
+        &mut self,
+        mut writer: Write,
+        input_index: usize,
+        script_code: &Script,
+        value: u64,
+        sighash_type: LegacySigHashType,
+    ) -> Result<(), Error> {
+        self.check_index(input_index)?;
+
+        let zero_hash = sha256d::Hash::default();
+
+        let (sighash, anyone_can_pay) = sighash_type.split_anyonecanpay_flag();
+
+        self.tx.version.consensus_encode(&mut writer)?;
+
+        if !anyone_can_pay {
+            self.segwit_cache().prevouts.consensus_encode(&mut writer)?;
+        } else {
+            zero_hash.consensus_encode(&mut writer)?;
+        }
+
+        if !anyone_can_pay
+            && sighash != LegacySigHashType::Single
+            && sighash != LegacySigHashType::None
+        {
+            self.segwit_cache()
+                .sequences
+                .consensus_encode(&mut writer)?;
+        } else {
+            zero_hash.consensus_encode(&mut writer)?;
+        }
+
+        {
+            let txin = &self.tx.input[input_index];
+
+            txin.previous_output.consensus_encode(&mut writer)?;
+            script_code.consensus_encode(&mut writer)?;
+            value.consensus_encode(&mut writer)?;
+            txin.sequence.consensus_encode(&mut writer)?;
+        }
+
+        if sighash != LegacySigHashType::Single && sighash != LegacySigHashType::None {
+            self.segwit_cache().outputs.consensus_encode(&mut writer)?;
+        } else if sighash == LegacySigHashType::Single && input_index < self.tx.output.len() {
+            let mut single_enc = SigHash::engine();
+            self.tx.output[input_index].consensus_encode(&mut single_enc)?;
+            SigHash::from_engine(single_enc).consensus_encode(&mut writer)?;
```

Comment:
Should add an `else if sighash == LegacySigHashType::Single && input_index >= self.tx.output.len()` which encodes the one hash

Reply:
Never mind.